In this notebook we will explore thed data to find ways to generate more revenue for taxi cab drivers. 
Our model wil Identify whether a customers is a generous tipper 


## Feature Engineering

In [1]:
#  Imports 
import pandas as pd
import numpy as np  

# Visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Data Preprocessing
from sklearn.preprocessing import StandardScaler, MinMaxScaler

# Model Selection 
from sklearn.model_selection import GridSearchCV, train_test_split

# Machine Learning Models
from sklearn.ensemble import RandomForestClassifier
from xgboost import XGBClassifier , plot_importance

# Evaluation Metrics (classification)
from sklearn.metrics import roc_auc_score, roc_curve
from sklearn.metrics import accuracy_score, precision_score, recall_score,\
f1_score, confusion_matrix, ConfusionMatrixDisplay, RocCurveDisplay

# Utilities
import warnings
warnings.filterwarnings("ignore")


In [5]:
pd.set_option('display.max_columns', None)

In [8]:
df0 = pd.read_csv('2017_Yellow_Taxi_Trip_Data.csv')
# NYC taxi fare predictions from the multiple linear regression model
nyc_preds_means = pd.read_csv('nyc_taxi_fare_predictions_mlr.csv')

In [9]:
nyc_preds_means.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   mean_duration   22699 non-null  float64
 1   mean_distance   22699 non-null  float64
 2   predicted_fare  22699 non-null  float64
dtypes: float64(3)
memory usage: 532.1 KB


In [10]:
df0.head()
# print(df0.shape)
# print(nyc_preds_means.shape)
#print(df0.index)
#print(nyc_preds_means.index)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  
0        2.76           0.0                    0.3         16.56  
1        4.00           0.0                    0.3         20.80  
2        1.45           0.0                    0.3          8.75  
3        6.39           0.0                    0.3         27.69  
4        0.00           0.0                    0.3         17.80

In [11]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22699 entries, 0 to 22698
Data columns (total 18 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             22699 non-null  int64  
 1   VendorID               22699 non-null  int64  
 2   tpep_pickup_datetime   22699 non-null  object 
 3   tpep_dropoff_datetime  22699 non-null  object 
 4   passenger_count        22699 non-null  int64  
 5   trip_distance          22699 non-null  float64
 6   RatecodeID             22699 non-null  int64  
 7   store_and_fwd_flag     22699 non-null  object 
 8   PULocationID           22699 non-null  int64  
 9   DOLocationID           22699 non-null  int64  
 10  payment_type           22699 non-null  int64  
 11  fare_amount            22699 non-null  float64
 12  extra                  22699 non-null  float64
 13  mta_tax                22699 non-null  float64
 14  tip_amount             22699 non-null  float64
 15  to

In [12]:
nyc_preds_means.head()

mean_duration  mean_distance  predicted_fare
0      22.847222       3.521667       16.434245
1      24.470370       3.108889       16.052218
2       7.250000       0.881429        7.053706
3      30.250000       3.700000       18.731650
4      14.616667       4.435000       15.845642

In [13]:
#merge the two dataframes on the index
df0 = df0.merge(nyc_preds_means, left_index=True, right_index=True)


In [14]:
df0.head(5)

Unnamed: 0  VendorID    tpep_pickup_datetime   tpep_dropoff_datetime  \
0    24870114         2   03/25/2017 8:55:43 AM   03/25/2017 9:09:47 AM   
1    35634249         1   04/11/2017 2:53:28 PM   04/11/2017 3:19:58 PM   
2   106203690         1   12/15/2017 7:26:56 AM   12/15/2017 7:34:08 AM   
3    38942136         2   05/07/2017 1:17:59 PM   05/07/2017 1:48:14 PM   
4    30841670         2  04/15/2017 11:32:20 PM  04/15/2017 11:49:03 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
4                1           4.37           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
4             4           112             2         16.5    0.5      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        2.76           0.0                    0.3         16.56   
1        4.00           0.0                    0.3         20.80   
2        1.45           0.0                    0.3          8.75   
3        6.39           0.0                    0.3         27.69   
4        0.00           0.0                    0.3         17.80   

   mean_duration  mean_distance  predicted_fare  
0      22.847222       3.521667       16.434245  
1      24.470370       3.108889       16.052218  
2       7.250000       0.881429        7.053706  
3      30.250000       3.700000       18.731650  
4      14.616667       4.435000       15.845642

### Feature Engineering

We are trying to predict if a customer will be a generous tipper.

Here are the codes in the `payment_type` column:

| Code | Payment Type     |
|------|------------------|
| 0    | Flex Fare trip   |
| 1    | Credit card      |
| 2    | Cash             |
| 3    | No charge        |
| 4    | Dispute          |
| 5    | Unknown          |
| 6    | Voided trip      |


In [15]:
df0['payment_type'].value_counts()

payment_type
1    15265
2     7267
3      121
4       46
Name: count, dtype: int64

Average tip of customers that pay cash

In [16]:
# Calculate the mean tip amount of payments in cash ( payment_type=2)
mean_tip_cash = df0[df0['payment_type'] == 2]['tip_amount'].mean()
print(f"Mean tip amount for cash payments: {mean_tip_cash:.2f}")

Mean tip amount for cash payments: 0.00


Taxi drivers do not report the tips received from customers that pay cash.  In the data the tip for the 7,267 trips paid with cash is zero.
For this study I will consider only customers that pay with credit card. 

In [17]:
#create a dataframe with only the credit card payments  
df1 = df0[df0['payment_type'] == 1].copy()

In [19]:
df1['tip_percent'] = round(df1['tip_amount'] / (df1['total_amount']- df1['tip_amount']),3)

In [20]:
df1.head(5)

Unnamed: 0  VendorID   tpep_pickup_datetime  tpep_dropoff_datetime  \
0    24870114         2  03/25/2017 8:55:43 AM  03/25/2017 9:09:47 AM   
1    35634249         1  04/11/2017 2:53:28 PM  04/11/2017 3:19:58 PM   
2   106203690         1  12/15/2017 7:26:56 AM  12/15/2017 7:34:08 AM   
3    38942136         2  05/07/2017 1:17:59 PM  05/07/2017 1:48:14 PM   
5    23345809         2  03/25/2017 8:34:11 PM  03/25/2017 8:42:11 PM   

   passenger_count  trip_distance  RatecodeID store_and_fwd_flag  \
0                6           3.34           1                  N   
1                1           1.80           1                  N   
2                1           1.00           1                  N   
3                1           3.70           1                  N   
5                6           2.30           1                  N   

   PULocationID  DOLocationID  payment_type  fare_amount  extra  mta_tax  \
0           100           231             1         13.0    0.0      0.5   
1           186            43             1         16.0    0.0      0.5   
2           262           236             1          6.5    0.0      0.5   
3           188            97             1         20.5    0.0      0.5   
5           161           236             1          9.0    0.5      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        2.76           0.0                    0.3         16.56   
1        4.00           0.0                    0.3         20.80   
2        1.45           0.0                    0.3          8.75   
3        6.39           0.0                    0.3         27.69   
5        2.06           0.0                    0.3         12.36   

   mean_duration  mean_distance  predicted_fare  tip_percent  
0      22.847222       3.521667       16.434245        0.200  
1      24.470370       3.108889       16.052218        0.238  
2       7.250000       0.881429        7.053706        0.199  
3      30.250000       3.700000       18.731650        0.300  
5      11.855376       2.052258       10.441351        0.200

We consider a generous tipper the customer that leaves a 20% tip or more.  Our target variable is "is a generos tipper or not "

In [ ]:
df1['generous'] = df1['tip_percent'] >= 0.20

#df1['generous'] = df1['generous'].astype(int)